File: scala/src/test/scala/NumericSpec.scala
import org.scalatest._
import collection.mutable.Seq

import algorithm._

class NumericSpec extends FlatSpec with Matchers {

  def xPlusY = (x: Int, y: Int) => x + y

  trait Fixture {
    var a = Seq(1, 2, 3)
    var b = Seq(4, 5, 6)
  }

  "Numeric Algorithms" should "Itoa OK" in new Fixture {
    val itoa = new PlusOne(4)
    Numeric.Itoa(a, itoa)
    a should be(b)
  }

  it should "Accumulate OK" in new Fixture {
    Numeric.Accumulate(a, 2, xPlusY) should be(8)
  }

  it should "InnerProduct OK" in new Fixture {
    val c = Seq(5, 7, 9)
    Numeric.InnerProduct(a, b, xPlusY) should be(c)
  }

  it should "AdjacentOp OK" in new Fixture {
    val c = Seq(1, 3, 5)
    Numeric.AdjacentOp(a, xPlusY) should be(c)
  }

  it should "PartialReduce OK" in new Fixture {
    val c = Seq(4, 9, 15)
    Numeric.PartialReduce(b, xPlusY) should be(c)
  }
}

File: typescript/test/numericSpec.js
var should = require('chai').should();
var Algos = require('../dist/numeric');

var Ary = function () {
  return [1, 2, 3];
};

describe('numeric algorithms', function(){

  it('itoa', function(){
    var g =  Algos.itoa(3);
    g().should.be.equal(3);
    g().should.be.equal(4);
  });

  it('accumulate', function(){
    Algos.accumulate([], 1, function(x, y) {
      return x * y;
    }).should.be.equal(1);
    Algos.accumulate(Ary(), 1, function(x, y) {
      return x * y;
    }).should.be.equal(6);
  });

  it('innerProduction', function(){
    Algos.innerProduction([], [], 0).should.be.equal(0);
    Algos.innerProduction(Ary(), Ary(), 0).should.be.equal(14);
  });

  it('adjacentDifference', function(){
    JSON.stringify(Algos.adjacentDifference([]))
      .should.be.equal(JSON.stringify([]));
    JSON.stringify(Algos.adjacentDifference(Ary()))
      .should.be.equal(JSON.stringify([1, 1, 1]));
  });

  it('partialSum', function(){
    JSON.stringify(Algos.partialSum([]))
      .should.be.equal(JSON.stringify([]));
    JSON.stringify(Algos.partialSum(Ary()))
      .should.be.equal(JSON.stringify([1, 3, 6]));
  });
});
